src/ClientTransactionTask.h

```diff
@@ -0,0 +1,264 @@
+
+        /// Copy constructor for CacheEntry, used to get around the missing
+        /// emplace feature in std::multimap.
+        explicit CacheEntry(const CacheEntry& other)
+            : type(other.type)
+            , objectBuf(other.objectBuf)
+            , rejectRules(other.rejectRules)
+            , rpcId(other.rpcId)
+            , state(other.state)
+        {}
+
+        /// Destructor for CacheEntry.
+        ///
+        /// Warning: Multiple copies of CacheEntry objects may cause the
+        /// ObjectBuffer pointed in the entry to be double freed.  This
+        /// is indirectly due to missing emplace feature in std::multimap.
+        ~CacheEntry()
+        {
+            if (objectBuf)
+                delete objectBuf;
+        }
+
+        /// Assignment operator for CacheEntry, used to get around the missing
+        /// emplace feature in std::multimap.
+        CacheEntry& operator=(const CacheEntry& other)
+        {
+            if (this != &other) {
+                type = other.type;
+                objectBuf = other.objectBuf;
+                rejectRules = other.rejectRules;
+                rpcId = other.rpcId;
+                state = other.state;
+            }
+            return *this;
+        }
+    };
+
+    explicit ClientTransactionTask(RamCloud* ramcloud);
+
+    CacheEntry* findCacheEntry(Key& key);
+    WireFormat::TxDecision::Decision getDecision();
+    Status getStatus() { return status; }
+    CacheEntry* insertCacheEntry(uint64_t tableId, const void* key,
+            uint16_t keyLength, const void* buf, uint32_t length);
+    /// Check if the task as completed the commit protocol.
+    bool isReady() { return (state == DONE); }
+    void performTask();
+
+  PRIVATE:
+    // Forward declaration of RPCs
+    class PrepareRpc;
+    class DecisionRpc;
+
+    /// Overall client state information.
+    RamCloud* ramcloud;
+
+    /// Number of participant objects/operations.
+    uint32_t participantCount;
+    /// List of participant object identifiers.
+    Buffer participantList;
+
+    /// Keeps track of the task currently executing phase.
+    enum State { INIT, PREPARE, DECISION, DONE} state;
+
+    /// Status of the transaction.  Used to defer exceptions.
+    Status status;
+
+    /// This transaction's decision to either COMMIT or ABORT.
+    WireFormat::TxDecision::Decision decision;
+
+    /// Lease information for to this transaction.
+    WireFormat::ClientLease lease;
+
+    /// Id of the rpcId that should be completed once the transaction is
+    /// complete.
+    uint64_t txId;
+
+    /// List of "in flight" Prepare Rpcs.
+    std::list<PrepareRpc> prepareRpcs;
+    /// List of "in flight" Decision Rpcs.
+    std::list<DecisionRpc> decisionRpcs;
+
+    /**
+     * Structure to define the key search value for the CommitCache map.
+     */
+    struct CacheKey {
+        uint64_t tableId;       // tableId of the tablet
+        KeyHash keyHash;        // start key hash value
```

Comment:
Is it possible to have multiple cache entries with the same keyHash? Let's discuss.